### PR TITLE
Added VS Version input to support back compatibility with VsBuild

### PR DESCRIPTION
**Pull request by t-dedah**
**Task name**: UseDotNetV2

**Description**: Added VS version as optional input which will be used to select the correct Dotnet SDK version for that VS version. This change is needed for customers using the VsBuild task rather than dotnetcore CLI task.

**Documentation changes required:** No

**Added unit tests:** : Updated necessary test cases to use the updated version of functions

**Attached related issue:** (Y/N) <Please add link to related issue here>

**Checklist**:
- [x] Task version was bumped - please check [instruction](https://github.com/microsoft/azure-pipelines-tasks/tree/master/docs/taskversionbumping.md) how to do it
- [ ] Checked that applied changes work as expected

### CODE REVIEW

File: Tasks/UseDotNetV2/README.md [hunk 1]
```diff
@@ -36,6 +36,8 @@ Examples:
   - To install latest patch version of 2.1 sdk, use 2.1.x
   - To install latest minor version of 2. sdk, use 2.x<br/>For getting more details about exact version, refer [this link](https://dotnetcli.blob.core.windows.net/dotnet/release-metadata/releases-index.json).
 
+* **Compatible Visual Studio Version**: Specify version of compatible visual studio for which .NET core sdk to install. Specifiy complete vs-version like 16.6.4 containing major version, minor version and patch number. Find the value of `vs-version` for installing SDK/Runtime, from the releases.json. The link to releases.json of that major.minor version can be found in [**releases-index file.**](https://dotnetcli.blob.core.windows.net/dotnet/release-metadata/releases-index.json)
```
Comment (bishal-pdMSFT) [inline Tasks/UseDotNetV2/README.md:39]: `Specify a compatible Visual Studio version. If a version is specified, then only those versions of .Net Core SDK/runtime will be considered for installation which are compatible with this Visual Studio version. Specify complete Visual Studio version containing major version, minor version and patch number. e.g. 16.6.4.
Compatibility of Visual Studio is checked by looking at `vs-version` property specified in the release manifest from the releases.json files. The link to releases.json of that major.minor version can be found in [**releases-index file.**](https://dotnetcli.blob.core.windows.net/dotnet/release-metadata/releases-index.json)`

File: Tasks/UseDotNetV2/versionfetcher.ts [hunk 2]
```diff
@@ -167,7 +167,11 @@ export class DotNetCoreVersionFetcher {
                             try {
                                 release.sdks.forEach((sdk) => {
                                     let versionInfo: VersionInfo = new VersionInfo(sdk, packageType);
-                                    versionInfoList.push(versionInfo);
+
+                                    if (!versionInfo.getvsVersion() || (vsVersionSpec && vsVersionSpec == versionInfo.getvsVersion()) || !vsVersionSpec) {
```
Comment (bishal-pdMSFT) [inline Tasks/UseDotNetV2/versionfetcher.ts:171]:
```suggestion
                                    if (!versionInfo.getvsVersion() || !vsVersionSpec || (vsVersionSpec == versionInfo.getvsVersion())) {
```

File: Tasks/UseDotNetV2/versionfetcher.ts [hunk 3]
```diff
@@ -177,7 +181,10 @@ export class DotNetCoreVersionFetcher {
                         if (release && release[packageType] && release[packageType].version && !versionInfoList.find((versionInfo) => { return versionInfo.getVersion() === release[packageType].version })) {
                             try {
                                 let versionInfo: VersionInfo = new VersionInfo(release[packageType], packageType);
-                                versionInfoList.push(versionInfo);
+
+                                if (!versionInfo.getvsVersion() || (vsVersionSpec && vsVersionSpec == versionInfo.getvsVersion()) || !vsVersionSpec) {
```
Comment (bishal-pdMSFT) [inline Tasks/UseDotNetV2/versionfetcher.ts:185]: same here

File: Tasks/UseDotNetV2/task.json [hunk 4]
```diff
@@ -70,6 +67,15 @@
...
             "visibleRule": "useGlobalJson = false || packageType = runtime"
         },
+        {
+            "name": "vsVersion",
+            "type": "string",
+            "label": "Compatible Visual Studio version",
+            "defaultValue": "",
+            "required": false,
+            "helpMarkDown": "Specify version of compatible visual studio for which .NET core sdk to install. Specifiy complete vs-version like 16.6.4 containing major version, minor version and patch number.Find the value of `version` for installing SDK/Runtime, from the releases.json. The link to releases.json of that major.minor version can be found in [**releases-index file.**](https://dotnetcli.blob.core.windows.net/dotnet/release-metadata/releases-index.json)",
+            "visibleRule": "useGlobalJson = false || packageType = runtime"
```
Comment (bishal-pdMSFT) [inline Tasks/UseDotNetV2/task.json:77]: why do only for runtime?

Comment (t-dedah) [reply Tasks/UseDotNetV2/task.json:77]: This is the same visible rule for vsVersion, I thought these two are linked together.

File: Tasks/UseDotNetV2/task.json [hunk 5]
```diff
@@ -70,6 +67,15 @@
             "helpMarkDown": "Specify version of .NET Core SDK or runtime to install.<br/>Versions can be given in the following formats<li>2.x   => Install latest in major version.</li><li>2.2.x => Install latest in major and minor version</li><li>2.2.104 => Install exact version</li><br/>Find the value of `version` for installing SDK/Runtime, from the releases.json. The link to releases.json of that major.minor version can be found in [**releases-index file.**](https://dotnetcli.blob.core.windows.net/dotnet/release-metadata/releases-index.json). Like link to releases.json for 2.2 version is https://dotnetcli.blob.core.windows.net/dotnet/release-metadata/2.2/releases.json",
             "visibleRule": "useGlobalJson = false || packageType = runtime"
         },
+        {
+            "name": "vsVersion",
```
Comment (bishal-pdMSFT) [inline Tasks/UseDotNetV2/task.json:71]: we should keep this input in advanced section

Comment (bishal-pdMSFT) [inline Tasks/UseDotNetV2/task.json:71]: we should update two documentations:
1. Update the readme in this doc to add this input
2. We should add a section for troubleshooting with the error people faced, and then suggest using the VS version as a fix. See [this](https://docs.microsoft.com/en-us/azure/devops/pipelines/tasks/deploy/azure-resource-group-deployment?view=azure-devops#troubleshooting) for reference

Comment (bishal-pdMSFT) [reply Tasks/UseDotNetV2/task.json:71]: For second one, it will not part of this PR. Talk to Nikitha (PM) with section content.

**Review by bishal-pdMSFT — APPROVED** [review]

**Pull request merged**